### ISSUE TITLE
Unexpected behaviour of the stubbing API (when using the API in a wrong or unintended way)

### ISSUE THREAD

**Issue by maxgrabenhorst** [mockito/mockito#1547]
### Problem
When using the stubbing API in an unintended way, this could lead to the following issues:

```
public class InvalidUsageTest extends TestBase {

    @Mock private IMethods mock;

    @Test
    public void failure_scenario_1() {
        mock.simpleMethod();
        final OngoingStubbing<String> ongoingStubbing = when(mock.simpleMethod());
        ongoingStubbing.thenReturn("A");
        ongoingStubbing.thenReturn("B");
        assertEquals("B", mock.simpleMethod());
    }

    @Test
    public void failure_scenario_2_1() {
        final OngoingStubbing<String> ongoingStubbing = when(mock.simpleMethod());
        ongoingStubbing.thenReturn("A");
        mock.differentMethod();
        ongoingStubbing.thenReturn("B");
        assertEquals("A", mock.simpleMethod());
        assertEquals("B", mock.differentMethod());
    }

    @Test
    public void failure_scenario_2_2() {
        final OngoingStubbing<String> ongoingStubbing =
                                when(mock.simpleMethod()).thenReturn("A");
        when(mock.differentMethod()).thenReturn("B");
        ongoingStubbing.thenReturn("C");
        assertEquals("A", mock.simpleMethod());
        assertNotEquals("C", mock.simpleMethod());
        assertEquals("B", mock.differentMethod());
        assertEquals("C", mock.differentMethod());
    }
}
```

#### Scenario 1
When calling a mock method before actual stubbing, it is possible to call the `thenReturn()` multiple times (as often as the mock method was called before). An exception would be expected but instead the first call of `thenReturn("A")` will be overridden with `thenReturn("B")`

#### Scenario 2.1
When calling a mock method (e.g. `mock.differentMethod()`)  between two calls of `.thenReturn()` on the same `OngoingStubbing` object (created by passing `mock.simpleMethod()` to `when` method), instead of getting an exception (or overriding the previous call, like scenario 1.1) the `differentMethod()` method will be stubbed.

#### Scenario 2.2
Quite the same as scenario 2.1, but occurs when stubbing consecutively.

#### Scenario 2.1 / 2.2
This happens because the behaviour of `OngoingStubbing` depends on the state of the `InvocationContainerImpl`. This state changes in the background while executing methods of the mock, without notifying the `OngoingStubbing` implementations about changes, e.g. change of the `invocationForStubbing`. When adding a new answer using an already created `OngoingStubbing`, it may happen that the `invocationForStubbing` has changed and hence the wrong method will be stubbed.
#### Scenario 1
In this case the check whether `thenReturn()` can be called relies on the `registeredInvocations` of the `InvocationContainerImpl` (when register not empty, `thenReturn()` can be called; `thenReturn()` removes last invocation from register). Here again this state can be modified by calling any method of the mock and hence it is possible to increase the `registeredInvocations` count before stubbing the actual method, which results in the possibility to call the `thenReturn()` method multiple times.

-------

What do you think Szczepan, should this unintended / wrong API usage result in proper exception messages or is this unexpected behaviour okay, when using the API in an unintended way?

**Comment by mockitoguy** [comment]
Thank you for reporting and taking the time to do thorough analysis.

>What do you think Szczepan, should this unintended / wrong API usage result in proper exception messages or is this unexpected behaviour okay, when using the API in an unintended way?

Good question! We want to strike a balance between "ease of use" and "internal complexity". We do want excellent exception messages for common pitfalls. We don't necessarily need great exceptions for "theoretical" pitfalls, especially when handling those scenarios lead to more complexity.

How did you come across those scenarios?

**Comment by maxgrabenhorst** [comment]
You are right. These are mainly just "theoretical" pitfalls. Your explanation makes absolutely sense to me and that's the reason why I'm asking before submitting a PR. Finding the right balance is important and it is not possible (or rather should be the goal) to prevent every misuse. I came across those scenarios when I was working on my last PR (#1541). While analysing the code I found out that these scenarios could happen.

For scenario 1 it would be necessary to keep the state if the `thenReturn` method was called in the `OngoingStubbingImpl` and check that state rather asking the `InvocationContainerImpl` if there is an invocation for stubbing (`hasInvocationForPotentialStubbing`). For scenario 2 we could add a listener to the `InvocationContainerImpl` notifying about the change of the `invocationForStubbing` and using this mechanism to invalidate the existing `OngoingStubbing` implementations.
When doing this the `hasInvocationForPotentialStubbing` method and the related tests could be deleted.

If you think Mockito should have exceptions for these pitfalls, I'm always happy to help 😃
